File: src/bindings/fixtures/crypto.lua
function base64()
  local crypto = require("@lmb/crypto")

  local text = " "

  local encoded = crypto.base64_encode(text)
  local expected = "IA=="
  assert(expected == encoded, "Expected '" .. expected .. "' but got '" .. encoded .. "'")
  local decoded = crypto.base64_decode(encoded)
  assert(text == decoded, "Expected '" .. text .. "' but got '" .. decoded .. "'")

  local crc32 = crypto.crc32(text)
  local expected = "e96ccf45"
  assert(expected == crc32, "Expected '" .. expected .. "' but got '" .. crc32 .. "'")
end

function hash()
  local crypto = require("@lmb/crypto")

  local text = " "

  local md5 = crypto.md5(text)
  local expected = "7215ee9c7d9dc229d2921a40e899ec5f"
  assert(expected == md5, "Expected '" .. expected .. "' but got '" .. md5 .. "'")

  local sha1 = crypto.sha1(text)
  local expected = "b858cb282617fb0956d960215c8e84d1ccf909c6"
  assert(expected == sha1, "Expected '" .. expected .. "' but got '" .. sha1 .. "'")

  local sha256 = crypto.sha256(text)
  local expected = "36a9e7f1c95b82ffb99743e0c5c4ce95d83c9a430aac59f84ef3cbfab6145068"
  assert(expected == sha256, "Expected '" .. expected .. "' but got '" .. sha256 .. "'")

  local sha384 = crypto.sha384(text)
  local expected = "588016eb10045dd85834d67d187d6b97858f38c58c690320c4a64e0c2f92eebd9f1bd74de256e8268815905159449566"
  assert(expected == sha384, "Expected '" .. expected .. "' but got '" .. sha384 .. "'")

  local sha512 = crypto.sha512(text)
  local expected =
    "f90ddd77e400dfe6a3fcf479b00b1ee29e7015c5bb8cd70f5f15b4886cc339275ff553fc8a053f8ddc7324f45168cffaf81f8c3ac93996f6536eef38e5e40768"
  assert(expected == sha512, "Expected '" .. expected .. "' but got '" .. sha512 .. "'")
end

function hmac_hash()
  local crypto = require("@lmb/crypto")

  local text = " "
  local key = "secret"

  local hmac_sha1 = crypto.hmac("sha1", text, key)
  local expected = "3fc26947ece0e3400c2216d2bcad669347e691ae"
  assert(expected == hmac_sha1, "Expected '" .. expected .. "' but got '" .. hmac_sha1 .. "'")

  local hmac_sha256 = crypto.hmac("sha256", text, key)
  local expected = "449cae45786ff49422f05eb94182fb6456b10db5c54f2342387168702e4f5197"
  assert(expected == hmac_sha256, "Expected '" .. expected .. "' but got '" .. hmac_sha256 .. "'")

  local hmac_sha384 = crypto.hmac("sha384", text, key)
  local expected = "82bb1e20f2d5c3ea86a7ecde4470923bfc7901b88a8154fe8e9ae9c326b822eb326b16517c72ee83294f376f6395a1ad"
  assert(expected == hmac_sha384, "Expected '" .. expected .. "' but got '" .. hmac_sha384 .. "'")

  local hmac_sha512 = crypto.hmac("sha512", text, key)
  local expected =
    "750198b84504923b67e3774963773255f4300aa7b3eaddd6a2eabb837d04c7f2e949d68faf22861fd1b560e66f7513eda5a47139a990f1ff5df90aac167fe4ca"
  assert(expected == hmac_sha512, "Expected '" .. expected .. "' but got '" .. hmac_sha512 .. "'")
end

function encrpyt_decrypt()
  local crypto = require("@lmb/crypto")

  local text = " "
  local key = "0123456701234567"
  local iv = "0123456701234567"

  local encrypted = crypto.encrypt("aes-cbc", text, key, iv)
  local expected = "b019fc0029f1ae88e96597dc0667e7c8"
  assert(expected == encrypted, "Expected '" .. expected .. "' but got '" .. encrypted .. "'")

  local decrypted = crypto.decrypt("aes-cbc", encrypted, key, iv)
  assert(text == decrypted, "Expected '" .. text .. "' but got '" .. decrypted .. "'")

  local key = "01234567"
  local iv = "01234567"
  local encrypted = crypto.encrypt("des-cbc", text, key, iv)
  local expected = "b865f90e7600d7ec"
  assert(expected == encrypted, "Expected '" .. expected .. "' but got '" .. encrypted .. "'")

  local decrypted = crypto.decrypt("des-cbc", encrypted, key, iv)
  assert(text == decrypted, "Expected '" .. text .. "' but got '" .. decrypted .. "'")

  local encrypted = crypto.encrypt("des-ecb", text, key)
  local expected = "7ab4f476973c70cc"
  assert(expected == encrypted, "Expected '" .. expected .. "' but got '" .. encrypted .. "'")

  local decrypted = crypto.decrypt("des-ecb", encrypted, key)
  assert(text == decrypted, "Expected '" .. text .. "' but got '" .. decrypted .. "'")
end

function crypto()
  base64()
  hash()
  hmac_hash()
  encrpyt_decrypt()
end

return crypto

File: src/bindings/crypto.rs
//! Cryptographic functions binding module.
//!
//! This module provides cryptographic utilities for hashing, encoding, and encryption.
//! Import via `require("@lmb/crypto")`.
//!
//! # Available Methods
//!
//! ## Encoding
//! - `base64_encode(data)` - Encode data to base64 string.
//! - `base64_decode(data)` - Decode base64 string to data.
//!
//! ## Hashing
//! - `crc32(data)` - Compute CRC32 checksum (hex string).
//! - `md5(data)` - Compute MD5 hash (hex string).
//! - `sha1(data)` - Compute SHA-1 hash (hex string).
//! - `sha256(data)` - Compute SHA-256 hash (hex string).
//! - `sha384(data)` - Compute SHA-384 hash (hex string).
//! - `sha512(data)` - Compute SHA-512 hash (hex string).
//! - `hmac(algorithm, data, secret)` - Compute HMAC (sha1, sha256, sha384, sha512).
//!
//! ## Encryption/Decryption
//! - `encrypt(cipher, data, key, iv)` - Encrypt data using specified cipher.
//! - `decrypt(cipher, encrypted, key, iv)` - Decrypt data using specified cipher.
//!
//! # Supported Ciphers
//!
//! | Cipher     | Key Size | IV Size | Notes                    |
//! |------------|----------|---------|--------------------------|
//! | `aes-cbc`  | 16 bytes | 16 bytes| AES-128 in CBC mode      |
//! | `des-cbc`  | 8 bytes  | 8 bytes | DES in CBC mode          |
//! | `des-ecb`  | 8 bytes  | N/A     | DES in ECB mode (no IV)  |
//!
//! # Security Warning
//!
//! - **DES is considered insecure** for modern applications. Use AES when possible.
//! - **MD5 and SHA-1** should not be used for security-critical hashing.
//! - Always use cryptographically secure random values for keys and IVs.
//!
//! # Example
//!
//! ```lua
//! local crypto = require("@lmb/crypto")
//!
//! -- Base64 encoding
//! local encoded = crypto.base64_encode("Hello")
//! local decoded = crypto.base64_decode(encoded)
//!
//! -- Hashing
//! local hash = crypto.sha256("password")
//!
//! -- HMAC
//! local hmac = crypto.hmac("sha256", "message", "secret")
//!
//! -- AES encryption (key and IV must be 16 bytes)
//! local key = "1234567890123456"
//! local iv = "abcdefghijklmnop"
//! local encrypted = crypto.encrypt("aes-cbc", "plaintext", key, iv)
//! local decrypted = crypto.decrypt("aes-cbc", encrypted, key, iv)
//! ```

use std::{fmt, sync::Arc};

use aes::cipher::{BlockDecryptMut as _, BlockEncryptMut as _, block_padding::Pkcs7};
use base64::prelude::*;
use crypto_common::{KeyInit, KeyIvInit as _};
use hmac::{Hmac, Mac};
use md5::Md5;
use mlua::prelude::*;
use sha1::Sha1;
use sha2::{Digest, Sha256, Sha384, Sha512};

type Aes128CbcEnc = cbc::Encryptor<aes::Aes128>;
type Aes128CbcDec = cbc::Decryptor<aes::Aes128>;
type DesCbcEnc = cbc::Encryptor<des::Des>;
type DesCbcDec = cbc::Decryptor<des::Des>;
type DesEcbEnc = ecb::Encryptor<des::Des>;
type DesEcbDec = ecb::Decryptor<des::Des>;

fn bad_argument(func: &str, pos: usize, message: impl fmt::Display) -> LuaError {
    LuaError::BadArgument {
        to: Some(func.to_string()),
        pos,
        name: None,
        cause: Arc::new(LuaError::external(message.to_string())),
    }
}

fn hash<H: Digest>(payload: &str) -> String {
    base16ct::lower::encode_string(&H::digest(payload))
}

fn hmac_hash<T: Mac + KeyInit>(secret: &str, payload: &str) -> mlua::Result<String> {
    let mut hasher = <T as KeyInit>::new_from_slice(secret.as_bytes()).into_lua_err()?;
    hasher.update(payload.as_bytes());
    let hash = hasher.finalize().into_bytes();
    Ok(base16ct::lower::encode_string(&hash))
}

pub(crate) struct CryptoBinding;

impl LuaUserData for CryptoBinding {
    fn add_methods<M: LuaUserDataMethods<Self>>(methods: &mut M) {
        methods.add_function("base64_encode", |_, data: String| {
            Ok(BASE64_STANDARD.encode(data.as_bytes()))
        });
        methods.add_function("base64_decode", |_, data: String| {
            let decoded = BASE64_STANDARD.decode(data.as_bytes()).into_lua_err()?;
            String::from_utf8(decoded).into_lua_err()
        });
        methods.add_function("crc32", |_, data: String| {
            Ok(format!("{:x}", crc32fast::hash(data.as_bytes())))
        });
        methods.add_function("md5", |_, data: String| Ok(hash::<Md5>(&data)));
        methods.add_function("sha1", |_, data: String| Ok(hash::<Sha1>(&data)));
        methods.add_function("sha256", |_, data: String| Ok(hash::<Sha256>(&data)));
        methods.add_function("sha384", |_, data: String| Ok(hash::<Sha384>(&data)));
        methods.add_function("sha512", |_, data: String| Ok(hash::<Sha512>(&data)));
        methods.add_function(
            "hmac",
            |_, (hash, data, secret): (String, String, String)| match hash.as_str() {
                "sha1" => hmac_hash::<Hmac<Sha1>>(&secret, &data),
                "sha256" => hmac_hash::<Hmac<Sha256>>(&secret, &data),
                "sha384" => hmac_hash::<Hmac<Sha384>>(&secret, &data),
                "sha512" => hmac_hash::<Hmac<Sha512>>(&secret, &data),
                _ => Err(bad_argument("hmac", 1, format!("unsupported hash {hash}"))),
            },
        );

        methods.add_function(
            "encrypt",
            |_, (cipher, data, key, iv): (String, String, String, Option<String>)| match cipher
                .as_str()
            {
                "aes-cbc" => {
                    let iv = iv.ok_or_else(|| bad_argument("encrypt", 4, "expect IV"))?;
                    let encrypted = Aes128CbcEnc::new(key.as_bytes().into(), iv.as_bytes().into())
                        .encrypt_padded_vec_mut::<Pkcs7>(data.as_bytes());
                    Ok(base16ct::lower::encode_string(&encrypted))
                }
                "des-cbc" => {
                    let iv = iv.ok_or_else(|| bad_argument("encrypt", 4, "expect IV"))?;
                    let encrypted = DesCbcEnc::new(key.as_bytes().into(), iv.as_bytes().into())
                        .encrypt_padded_vec_mut::<Pkcs7>(data.as_bytes());
                    Ok(base16ct::lower::encode_string(&encrypted))
                }
                "des-ecb" => {
                    let encrypted = DesEcbEnc::new(key.as_bytes().into())
                        .encrypt_padded_vec_mut::<Pkcs7>(data.as_bytes());
                    Ok(base16ct::lower::encode_string(&encrypted))
                }
                _ => Err(bad_argument(
                    "encrypt",
                    1,
                    format!("unsupported cipher {cipher}"),
                )),
            },
        );
        methods.add_function(
            "decrypt",
            |_, (cipher, encrypted, key, iv): (String, String, String, Option<String>)| match cipher
                .as_str()
            {
                "aes-cbc" => {
                    let iv = iv.ok_or_else(|| bad_argument("decrypt", 4, "expect IV"))?;
                    let data = hex::decode(&encrypted).into_lua_err()?;
                    let decrypted = Aes128CbcDec::new(key.as_bytes().into(), iv.as_bytes().into())
                        .decrypt_padded_vec_mut::<Pkcs7>(&data)
                        .map_err(|e| LuaError::external(e.to_string()))?;
                    Ok(String::from_utf8(decrypted).into_lua_err()?)
                }
                "des-cbc" => {
                    let iv = iv.ok_or_else(|| bad_argument("decrypt", 4, "expect IV"))?;
                    let data = hex::decode(&encrypted).into_lua_err()?;
                    let decrypted = DesCbcDec::new(key.as_bytes().into(), iv.as_bytes().into())
                        .decrypt_padded_vec_mut::<Pkcs7>(&data)
                        .map_err(|e| LuaError::external(e.to_string()))?;
                    Ok(String::from_utf8(decrypted).into_lua_err()?)
                }
                "des-ecb" => {
                    let data = hex::decode(&encrypted).into_lua_err()?;
                    let decrypted = DesEcbDec::new(key.as_bytes().into())
                        .decrypt_padded_vec_mut::<Pkcs7>(&data)
                        .map_err(|e| LuaError::external(e.to_string()))?;
                    Ok(String::from_utf8(decrypted).into_lua_err()?)
                }
                _ => Err(bad_argument(
                    "decrypt",
                    1,
                    format!("unsupported cipher {cipher}"),
                )),
            },
        );
    }
}

#[cfg(test)]
mod tests {
    use tokio::io::empty;

    use crate::Runner;

    #[tokio::test]
    async fn test_crypto() {
        let source = include_str!("../fixtures/bindings/crypto.lua");
        let runner = Runner::builder(source, empty()).build().unwrap();
        runner.invoke().call().await.unwrap().result.unwrap();
    }

    #[tokio::test]
    async fn test_crypto_errors() {
        let source = include_str!("../fixtures/bindings/crypto-errors.lua");
        let runner = Runner::builder(source, empty()).build().unwrap();
        runner.invoke().call().await.unwrap().result.unwrap();
    }

    #[tokio::test]
    async fn test_base64_roundtrip() {
        let source = include_str!("../fixtures/bindings/crypto/base64-roundtrip.lua");
        let runner = Runner::builder(source, empty()).build().unwrap();
        runner.invoke().call().await.unwrap().result.unwrap();
    }

    #[tokio::test]
    async fn test_hash_functions() {
        let source = include_str!("../fixtures/bindings/crypto/hash-functions.lua");
        let runner = Runner::builder(source, empty()).build().unwrap();
        runner.invoke().call().await.unwrap().result.unwrap();
    }

    #[tokio::test]
    async fn test_des_ecb_encryption() {
        let source = include_str!("../fixtures/bindings/crypto/des-ecb.lua");
        let runner = Runner::builder(source, empty()).build().unwrap();
        runner.invoke().call().await.unwrap().result.unwrap();
    }

    #[tokio::test]
    async fn test_des_cbc_encryption() {
        let source = include_str!("../fixtures/bindings/crypto/des-cbc.lua");
        let runner = Runner::builder(source, empty()).build().unwrap();
        runner.invoke().call().await.unwrap().result.unwrap();
    }
}
